dali/core/float16_test.cu

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <vector>
+#include "dali/core/float16.h"
+#include "dali/test/device_test.h"
+#include "dali/core/dev_buffer.h"
+
+namespace dali {
+
+TEST(FP16Host, Conversions) {
+  float16 a = {};
+#ifndef __CUDA_ARCH__
+  EXPECT_EQ(a.impl, float16::impl_t());
+#endif
+  a = 42;
+  EXPECT_EQ(static_cast<float>(a), 42.0f);
+  a = -42L;
+  EXPECT_EQ(static_cast<float>(a), -42);
+  long long ll = 1234;  // NOLINT
+  a = ll;
+  EXPECT_EQ(static_cast<float>(a), 1234);
+  a = -5.5f;
+  EXPECT_EQ(static_cast<float>(a), -5.5f);
+  a = 2u;
+  EXPECT_EQ(static_cast<float>(a), 2.0f);
+}
+
+DEVICE_TEST(FP16Dev, Conversions, 1, 1) {
```

Comment:
Can this by any chance be different when constructing values like
`float16 a = 42.f` (or `float16 a{42.f}`) rather than using assignment operators?

Can you check this, we had those really weird bugs around that area.

Also, what if I assign the `__half` or `half_float::half` types directly?

Reply:
Done.